### DISCUSSION TITLE
group->compute->pivot rearrangement for table definitions inspired by ggplot2 experience!

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#19] [Ideas]
I have this table project that reverses data processing steps to feel more like ggplot2. 

table creation can look like:

data |>
group_by(x, y) |>
summarize(means = mean(value)) |>
pivot_wider(cols = x) 

In ggplot2, describe table layout is first, and grouping is automatic

StatSum$default_aes <- aes(label = after_stat(n))

data |>
  ggplot(aes(x,y)) + 
  stat_sum(geom = "text")

This describe-the-layout contributes greatly to ggplot2 ease of use...  Favorite ever Hadley Wickham quotation: 

>  And, you know, I’d get a dataset. And, in my head I could very clearly kind of picture, I want to put this on the x-axis. Let’s put this on the y-axis, draw a line, put some points here, break it up by this variable. And then, like, getting that vision out of my head, and into reality, it’s just really, really hard. Just, like, felt harder than it should be. Like, there’s a lot of custom programming involved, where I just felt, like, to me, I just wanted to say, like, you know, this is what I’m thinking, this is how I’m picturing this plot. Like you’re the computer ‘Go and do it’. … and I’d also been reading about the Grammar of Graphics by Leland Wilkinson, I got to meet him a couple of times and … I was, like, this book has been, like, written for me. - https://www.trifacta.com/podcast/tidy-data-with-hadley-wickham/

I've worked on pivotr function to make the group by and pivot to be single be declarations which is the description of the layout of the table: 

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/bcdeb737-e32f-4c4f-857b-4af318ee5623)

So you can do these kind of moves: 

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/b393dda2-f18c-499a-910e-1e5aea1c5e05)

And then creating a few helpers to tell you about the computation to be done... 

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/3b7ad7e8-0e07-4629-85ab-bee18459b0c8)

But I didn't make progress for a long time on piping (allowing tables to evolve like plots with orthogonal declarations about the table), which a few people said was pretty desirable and I definitely see that.   @brshallo @shannonpileggi 
 https://github.com/EvaMaeRey/mytidytuesday/issues/3

I returned to it and it is very much proof of concept/OOP hack job. I use R6 but I didn't seem much on how to make user-facing functions, so I just hacked away.  It is ugly!!    https://evamaerey.github.io/mytidytuesday/2024-05-27-r6/r6.html  

The API is close to the target though.  It let's you build up a table: data; rows; cols
![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/9c905f11-e6b7-48fe-a91c-34775f42af9a)
 
Family of tables example with the last_table() is kind of cool and feels very ggplot2 imho
![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/65924a0b-7707-4a46-b5bd-0c5e76379919)

I learned about janitor::tabyl(data, rows = ?, cols = ?) in the early days, which is really really nice and I used a ton thereafter, but it's not as computationally flexible as you might like.  (It does auto add margins, which you also might like but are not in the immediate project scope - I think feeding the prepared table to something like gt would be perhaps *the* way)

Formally have presented the idea motivated by bright yellow hinged door so you know it's good. https://evamaerey.github.io/tidypivot/maa_may_2022_presentation.html#1  
![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/94a6b088-d663-4f4a-a9ff-08ce7613594f)

**Comment by teunbrand** [comment]
I like the syntax, I always keep forgetting the correct incantations to get these types of functions to do what I want

**Reply by EvaMaeRey** [reply]
Now in code-movie form https://evamaerey.github.io/mytidytuesday/2024-05-27-r6/flipbooks/tablebuildup.html#1 and table-contest submitted for reach: https://forum.posit.co/t/declarative-basic-pivot-table-construction-table-contest/188205   Wish I were a little further along in my OOP journey to support this better.

**Reply by EvaMaeRey** [reply]
putting @Cghlewis tweet here.  👋   https://twitter.com/Cghlewis/status/1803260895587348569  'It's 10pm and I'm looking up the arguments for [#rstats](https://twitter.com/hashtag/rstats?src=hashtag_click) pivot_wider() for the 100th time. 😅'   I'm the same.

**Reply by EvaMaeRey** [reply]
A compute_layer experiment shows some declarative in facet row-col actual ggplot2 table/heatmap. https://evamaerey.github.io/mytidytuesday/2024-06-03-table-stat-layer/table-stat-layer.html    Using lessons learned in https://github.com/teunbrand/ggplot-extension-club/discussions/16

**Comment by EvaMaeRey** [comment]
I decided to return to compute_layer exercise and do reprex here because of @yjunechoe #23 prop_by example.  Like @larmarange's StatProp, StatTableprop allows a  'prop_within' ('by') aes that defines the scope within which prop/percent should be calculated (else prop is computed on all data).

``` r
library(tidyverse)

Titanic %>% 
  data.frame() %>%  
  uncount(Freq) ->
tidy_titanic

compute_layer_prop_within <- function(data = tidy_titanic, ...){
  
  if(is.null(data$value)){data$value <- 1}
  if(is.null(data$prop_within)){data$prop_within <- 1}
  if(is.null(data$wt)){data$wt <- 1}
  
  data$value <- data$value * data$wt
  
  data %>% 
    group_by(prop_within) %>% 
    summarize(group_sum = sum(.data$value)) ->
  group_sum_df
  
  data %>% 
    left_join(group_sum_df, by = "prop_within") %>% 
    group_by(PANEL) %>% 
    summarize(sum = sum(.data$value),
              group_sum = mean(.data$group_sum)) %>% 
    mutate(prop = sum/group_sum) %>% 
    mutate(label = round(100 * prop, 1) %>% paste0("%")) %>% 
    mutate(x = 0, y = 0)

}


# test compute layer
tidy_titanic %>% 
  mutate(prop_by = Sex) %>% 
  mutate(PANEL = paste(as.numeric(Sex), 
                       as.numeric(Survived), sep = "-")) %>%
  ungroup() %>% 
  compute_layer_prop_within()
#> # A tibble: 4 × 7
#>   PANEL   sum group_sum   prop label     x     y
#>   <chr> <dbl>     <dbl>  <dbl> <chr> <dbl> <dbl>
#> 1 1-1    1364      2201 0.620  62%       0     0
#> 2 1-2     367      2201 0.167  16.7%     0     0
#> 3 2-1     126      2201 0.0572 5.7%      0     0
#> 4 2-2     344      2201 0.156  15.6%     0     0
```

``` r


StatTableprop <- ggproto(`_class` = "StatTableprop",
                    `_inherit` = Stat,
                    compute_layer = compute_layer_prop_within)

stat_tableprop <- function(geom = "text"){
  
    layer(geom = geom, 
        stat = StatTableprop, 
        position = "identity")
  
}

tidy_titanic %>% 
  ggplot() + 
  facet_wrap(~Sex) + 
  stat_tableprop()
```

![](https://i.imgur.com/V5jInVG.png)

``` r

last_plot() + 
  facet_grid(Sex ~ Survived)
```

![](https://i.imgur.com/DeHmIRm.png)

``` r

last_plot() + 
  aes(prop_within = Sex)
```

![](https://i.imgur.com/DtUjxF7.png)

``` r


last_plot() + 
  facet_grid(rows = vars(Sex),
             cols = vars(Age, Survived)) + 
  aes(prop_within = Sex)
```

![](https://i.imgur.com/uyb41GM.png)

``` r

and <- paste  # not super careful category combos


last_plot() +
  aes(prop_within = and(Survived, Age))
```

![](https://i.imgur.com/wK7PWwc.png)

``` r

layer_data()
#> # A tibble: 8 × 16
#>   PANEL   sum group_sum   prop label     x     y colour  size angle hjust vjust
#>   <fct> <dbl>     <dbl>  <dbl> <chr> <dbl> <dbl> <chr>  <dbl> <dbl> <dbl> <dbl>
#> 1 1        35        52 0.673  67.3%     0     0 black   3.88     0   0.5   0.5
#> 2 2        29        57 0.509  50.9%     0     0 black   3.88     0   0.5   0.5
#> 3 3      1329      1438 0.924  92.4%     0     0 black   3.88     0   0.5   0.5
#> 4 4       338       654 0.517  51.7%     0     0 black   3.88     0   0.5   0.5
#> 5 5        17        52 0.327  32.7%     0     0 black   3.88     0   0.5   0.5
#> 6 6        28        57 0.491  49.1%     0     0 black   3.88     0   0.5   0.5
#> 7 7       109      1438 0.0758 7.6%      0     0 black   3.88     0   0.5   0.5
#> 8 8       316       654 0.483  48.3%     0     0 black   3.88     0   0.5   0.5
#> # ℹ 4 more variables: alpha <lgl>, family <chr>, fontface <dbl>,
#> #   lineheight <dbl>
```

``` r

# implementation does not protect you from illogical specification. 
# (prop within both Survived and age, but table only breaks down by age.)
last_plot() + 
  facet_grid(rows = vars(Age))
```

![](https://i.imgur.com/B9eTAX0.png)

``` r

last_plot() +
  aes(prop_within = Age)
```

![](https://i.imgur.com/mzMuGsm.png)

``` r

last_plot() + 
  aes(prop_within = NULL)
```

![](https://i.imgur.com/5D4Olmk.png)

``` r

layer_data()
#> # A tibble: 2 × 16
#>   PANEL   sum group_sum   prop label     x     y colour  size angle hjust vjust
#>   <fct> <dbl>     <dbl>  <dbl> <chr> <dbl> <dbl> <chr>  <dbl> <dbl> <dbl> <dbl>
#> 1 1       109      2201 0.0495 5%        0     0 black   3.88     0   0.5   0.5
#> 2 2      2092      2201 0.950  95%       0     0 black   3.88     0   0.5   0.5
#> # ℹ 4 more variables: alpha <lgl>, family <chr>, fontface <dbl>,
#> #   lineheight <dbl>
```

<sup>Created on 2024-06-24 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Comment by EvaMaeRey** [comment]
Work in progress now in S3, 

https://evamaerey.github.io/mytidytuesday/2024-07-02-s3-tables/s3-tables.html#desired-usage-piped

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/2fe7bda9-42cf-4b28-9262-76576c02d369)

Not finding much material on S3 prepared for monolith function -> piped interface.  Maybe this could become that resource!  @ddsjoberg, @shannonpileggi actually recommended getting in touch about piping in this project long ago.   Maybe you could have a look or point to resources?

**Reply by EvaMaeRey** [reply]
Big picture ambition: 

![image](https://github.com/teunbrand/ggplot-extension-club/assets/10491910/e7991dc7-f3ab-4db4-a609-6b28e04338f4)


 (pivot_count is pruned back from tidypivot function which that lets you do more count, so after more than vanilla contingency table; just looking at count so that we can easily include complete code for unpiped version)

**Comment by EvaMaeRey** [comment]
Demonstration of more possible functionality.  

We incrementally build 1. a table that gives counts for observations (species X bird).  And then based on that layout specification (defaults to count), compute mean for the value column bill_depth_mm.

![image](https://github.com/user-attachments/assets/ee054200-91b1-4b7a-a2e3-bf62996d696c)

**Reply by EvaMaeRey** [reply]
for this demo, I just did remove_missing(penguins) -> penguins.  Missing data defaults is a big question.

**Reply by EvaMaeRey** [reply]
fun read on interaction w/ Pito Salas on pivot table tool.  https://blog-devresults-com.cdn.ampproject.org/v/s/blog.devresults.com/so-long-pito-old-friend/amp/?amp_gsa=1&amp_js_v=a9&usqp=mq331AQIUAKwASCAAgM%3D#amp_tf=From%20%251%24s&aoh=17277550622373&referrer=https%3A%2F%2Fwww.google.com&ampshare=https%3A%2F%2Fblog.devresults.com%2Fso-long-pito-old-friend%2F

**Reply by EvaMaeRey** [reply]
I've noticed the tabulation ambitions in ggsql by @teunbrand  https://github.com/posit-dev/ggsql/pull/546 ..  🙌🎉
I think it might be a good time to come back to this disucssion.

Loved your generous feedback @teunbrand then, 'I like the syntax, I always keep forgetting the correct incantations to get these types of functions to do what I want' 🙏🙂  

I haven't seen any ggsql syntax yet, but 'TABULATE statement, parallel to VISUALISE statements' makes it sound like the declarative table construction -- set rows, set cols, set aggregation --- is what's going on? (what I *tried* to get a handle on in this thread... 🤷‍♀️🫠)   Is that right?

**Reply by teunbrand** [reply]
Yeah it is very early days for me to be working on this, but I think the simplest explanation would be 'gt but for sql'. Though we're realigning some grammar to fit ggsql better. The reason we want to keep this together with the plot bits is that once we have plot composition going, you can go right ahead and combine plots with tables.

**Comment by EvaMaeRey** [comment]
I think this project will be motivating to work on if the there is a global option to print the tables with something like gt::gt or knitr::kable, though default should probably be to just print a tibble -- also good for report writing.  Will have a look at discussion on options() in #48

**Comment by EvaMaeRey** [comment]
One thing that make ggplot nice to use and fun is not having to create variables before you use them: aes(x = gdpPerCap * pop)  or aes(y = pop/1000000).  and facet_wrap(vars(continent == "Americas"))  This on-the-fly definition would be nice to have!
